src/core/int128.d

```diff
@@ -0,0 +1,748 @@
+enum MinusOne = neg(One);
+
+/*****************************
+ * Test against 0
+ * Params:
+ *      c = Cent to test
+ * Returns:
+ *      true if != 0
+ */
+pure
+bool tst(Cent c)
+{
+    return c.hi || c.lo;
+}
+
+
+/*****************************
+ * Complement
+ * Params:
+ *      c = Cent to complement
+ * Returns:
+ *      complemented value
+ */
+pure
+Cent com(Cent c)
+{
+    c.lo = ~c.lo;
+    c.hi = ~c.hi;
+    return c;
+}
+
+/*****************************
+ * Negate
+ * Params:
+ *      c = Cent to negate
+ * Returns:
+ *      negated value
+ */
+pure
+Cent neg(Cent c)
+{
+    return inc(com(c)); // ~c + 1
+}
+
+/*****************************
+ * Increment
+ * Params:
+ *      c = Cent to increment
+ * Returns:
+ *      incremented value
+ */
+pure
+Cent inc(Cent c)
+{
+    return add(c, One);
+}
+
+/*****************************
+ * Decrement
+ * Params:
+ *      c = Cent to decrement
+ * Returns:
+ *      incremented value
+ */
+pure
+Cent dec(Cent c)
+{
+    return sub(c, One);
+}
+
+/*****************************
+ * Shift left one bit
+ * Params:
+ *      c = Cent to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent shl1(Cent c)
+{
+    c.hi = (c.hi << 1) | (cast(I)c.lo < 0);
+    c.lo <<= 1;
+    return c;
+}
+
+/*****************************
+ * Unsigned shift right one bit
+ * Params:
+ *      c = Cent to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent shr1(Cent c)
+{
+    c.lo = (c.lo >> 1) | ((c.hi & 1) << (Ubits - 1));
+    c.hi >>= 1;
+    return c;
+}
+
+
+/*****************************
+ * Arithmetic shift right one bit
+ * Params:
+ *      c = Cent to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent sar1(Cent c)
+{
+    c.lo = (c.lo >> 1) | ((c.hi & 1) << (Ubits - 1));
+    c.hi = cast(I)c.hi >> 1;
+    return c;
+}
+
+/*****************************
+ * Shift left n bits
+ * Params:
+ *      c = Cent to shift
+ *      n = number of bits to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent shl(Cent c, uint n)
+{
+    if (n >= Ubits * 2)
+        return Zero;
+
+    if (n >= Ubits)
+    {
+        c.hi = c.lo << (n - Ubits);
+        c.lo = 0;
+    }
+    else
+    {
+        c.hi = ((c.hi << n) | (c.lo >> (Ubits - n - 1) >> 1));
+        c.lo = c.lo << n;
+    }
+    return c;
+}
+
+/*****************************
+ * Unsigned shift right n bits
+ * Params:
+ *      c = Cent to shift
+ *      n = number of bits to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent shr(Cent c, uint n)
+{
+    if (n >= Ubits * 2)
+        return Zero;
+
+    if (n >= Ubits)
+    {
+        c.lo = c.hi >> (n - Ubits);
+        c.hi = 0;
+    }
+    else
+    {
+        c.lo = ((c.lo >> n) | (c.hi << (Ubits - n - 1) << 1));
+        c.hi = c.hi >> n;
+    }
+    return c;
+}
+
+/*****************************
+ * Arithmetic shift right n bits
+ * Params:
+ *      c = Cent to shift
+ *      n = number of bits to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent sar(Cent c, uint n)
```

Comment:
It'd certainly improve usability if all these functions weren't named after x86 mnemonics.  For instance, `lshift`, `rshift`, and `arshift` instead of `shl`, `shr`, and `sar`.  Similarly, the rotate functions could be `lrotate` and `rrotate`, and the signed comparison functions be prefixed with `s` to make it clear that they are signed versions of the unsigned counterparts (eg: `sgt` and `ugt`).

Reply:
It doesn't exactly follow the x86 mnemonics, and I've seen those mnemonics more or less used by various CPUs my entire career. They're second nature :-)

I used a `u` prefix for unsigned, and no prefix for signed, to match D's byte/ubyte, short/ushort, cent/ucent, etc. I deliberately did not use the x86 mul/imul, div/idiv mnemonics.

Reply:
However, this is a public interface and so the naming conventions should really resemble something suitable for use outside of the compiler run-time.

Reply:
byte/ubyte, etc., are a public interface.